File: 833-bus-routes/bus-routes.py
from collections import defaultdict

class Solution:
    def numBusesToDestination(self, routes: List[List[int]], source: int, target: int) -> int:
        """
        1 -> 2 -> 7
        3 -> 6 -> 7

        """
        if source == target:
            return 0
        
        adj = [set() for _ in range(len(routes))] # bus_index -> bus index
        visited = [False] * len(routes)
        routes = [set(route) for route in routes]
        for i in range(len(routes)):
            for j in range(i + 1, len(routes)):
                if routes[i] & routes[j]:
                    adj[i].add(j)
                    adj[j].add(i)
        
        q = deque() # source buses
        for i, route in enumerate(routes):
            if source in route:
                if target in route:
                    return 1
                q.append(i)
        
        if not q:
            return -1

        # print(f"{adj} \n {q}")
        num_buses = 1

        while q: # run multi-source bfs
            for _ in range(len(q)):
                curr_bus = q.popleft()
                if visited[curr_bus]: continue

                visited[curr_bus] = True
                for next_bus in adj[curr_bus]:
                    if target in routes[next_bus]:
                        return num_buses + 1
                    if not visited[next_bus]:
                        q.append(next_bus)
                
            num_buses += 1
        
        return -1

File: 833-bus-routes/bus-routes.go
func numBusesToDestination(routes [][]int, source int, target int) int {
    if source == target {
        return 0
    }

    stopMap := make(map[int][]int, 0)
    for busNum, stops := range routes {
        for _, stop := range stops {
            stopMap[stop] = append(stopMap[stop], busNum)
        }

    }

    q := make([]int, 0) // q to store the busNum currently visited
    visited := make([]bool, len(routes)) // track with stops have been visited
    for _, busNum := range stopMap[source] {
        q = append(q, busNum)
        visited[busNum] = true
    }

    numBuses := 1 // have to start with 1 bus (soruce bus)

    for len(q) > 0 {
        currBuses := len(q) // number of buses at current level
        // fmt.Println("level", numBuses, q)

        for range currBuses {
            busNum := q[0]
            q = q[1:] // pop bus from front

            for _, stop := range routes[busNum] {
                if stop == target {
                    return numBuses
                }
                for _, nextBus := range stopMap[stop] {
                    if !visited[nextBus] {
                        q = append(q, nextBus)
                        visited[nextBus] = true
                    }
                }
            }
        }
        numBuses++
    }

    return -1;

}